File: hash/rust/rosetta-image-hash/tests/group3_test.rs
mod testkit;

use std::collections::HashSet;
use std::fs;
use std::io::{BufRead, BufReader};

use image::DynamicImage;
use rosetta_image_hash::{average_hash, colorhash, dhash, phash, whash_haar, Hash, ImageHashError};

fn load_exemptions() -> HashSet<String> {
	let mut exempt = HashSet::new();
	let Ok(f) = fs::File::open("DECODER_NOTES.md") else { return exempt };
	let reader = BufReader::new(f);
	for line in reader.lines().map_while(|r| r.ok()) {
		if let Some(idx) = line.find('—') {
			let name = line[..idx].trim();
			if name.ends_with(".png") {
				exempt.insert(name.to_string());
			}
		}
	}
	exempt
}

fn decode_png(fixture: &str) -> DynamicImage {
	let path = format!("{}/fixtures/{}", testkit::SPEC_DIR, fixture);
	image::open(&path).unwrap_or_else(|e| panic!("open {path}: {e}"))
}

fn run(
	name: &str,
	exempt: &HashSet<String>,
	compute: impl Fn(&DynamicImage, usize) -> Result<Hash, ImageHashError>,
	label: &str,
	failures: &mut Vec<String>,
) {
	for c in testkit::algorithm_cases(name) {
		if exempt.contains(&c.fixture) {
			continue;
		}
		let img = decode_png(&c.fixture);
		match compute(&img, c.size) {
			Err(e) => failures.push(format!("{label} fixture={} size={}: {e}", c.fixture, c.size)),
			Ok(h) => {
				if h.to_hex() != c.hex {
					failures.push(format!(
						"{label} fixture={} size={}: got {} want {}",
						c.fixture, c.size, h.to_hex(), c.hex
					));
				}
			}
		}
	}
}

#[test]
fn png_end_to_end() {
	let exempt = load_exemptions();
	let mut failures: Vec<String> = Vec::new();
	run("average_hash", &exempt, average_hash, "average_hash", &mut failures);
	run("dhash",        &exempt, dhash,        "dhash",        &mut failures);
	run("phash",        &exempt, phash,        "phash",        &mut failures);
	run("whash_haar",   &exempt, whash_haar,   "whash_haar",   &mut failures);
	run("colorhash",    &exempt, colorhash,    "colorhash",    &mut failures);
	if !failures.is_empty() {
		panic!("{} Group-3 failures:\n  {}", failures.len(), failures.join("\n  "));
	}
}

File: hash/js/rosetta-image-hash/tests/group3.test.ts
import { describe, it, expect } from "vitest";
import { readFileSync, existsSync } from "node:fs";
import { join } from "node:path";

import * as rih from "../src/index.js";
import { algorithmCases, SPEC_DIR } from "./testkit.js";
import type { Hash, RgbImage } from "../src/index.js";

function loadExemptions(): Set<string> {
	const exempt = new Set<string>();
	const path = "DECODER_NOTES.md";
	if (!existsSync(path)) return exempt;
	const text = readFileSync(path, "utf8");
	for (const line of text.split("\n")) {
		const idx = line.indexOf("—");
		if (idx > 0) {
			const name = line.slice(0, idx).trim();
			if (name.endsWith(".png")) exempt.add(name);
		}
	}
	return exempt;
}

function decodePngFromSpec(fixture: string): RgbImage {
	const path = join(SPEC_DIR, "fixtures", fixture);
	const bytes = new Uint8Array(readFileSync(path));
	return rih.decodePng(bytes);
}

function run(
	name: string,
	exempt: Set<string>,
	compute: (img: RgbImage, sz: number) => Hash,
	label: string,
	failures: string[],
) {
	for (const c of algorithmCases(name)) {
		if (exempt.has(c.fixture)) continue;
		const img = decodePngFromSpec(c.fixture);
		const h = compute(img, c.size);
		if (h.toHex() !== c.hex) {
			failures.push(`${label} fixture=${c.fixture} size=${c.size}: got ${h.toHex()} want ${c.hex}`);
		}
	}
}

describe("PNG end-to-end (Group 3)", () => {
	it("byte-exact for all algorithms × fixtures via decodePng", () => {
		const exempt = loadExemptions();
		const failures: string[] = [];
		run("average_hash", exempt, rih.averageHash, "average_hash", failures);
		run("dhash", exempt, rih.dhash, "dhash", failures);
		run("dhash_vertical", exempt, rih.dhashVertical, "dhash_vertical", failures);
		run("phash", exempt, rih.phash, "phash", failures);
		run("phash_simple", exempt, rih.phashSimple, "phash_simple", failures);
		run("whash_haar", exempt, rih.whashHaar, "whash_haar", failures);
		run("whash_db4", exempt, rih.whashDb4, "whash_db4", failures);
		run("colorhash", exempt, rih.colorhash, "colorhash", failures);
		if (failures.length > 0) {
			throw new Error(`${failures.length} Group-3 failures:\n  ${failures.join("\n  ")}`);
		}
	});
});
